File: downloader.py
"""
Rongyok Video Downloader
Downloads videos with resume support and progress tracking
"""

import os
import json
import time
import requests
from pathlib import Path
from typing import Optional, Callable
from dataclasses import dataclass, asdict
from tqdm import tqdm

from parser import EpisodeInfo, SeriesInfo


@dataclass
class DownloadProgress:
    """Track download progress for resume"""
    episode: int
    downloaded_bytes: int
    total_bytes: int
    completed: bool = False


@dataclass
class DownloadState:
    """Overall download state for a series"""
    series_id: int
    series_title: str
    total_episodes: int
    output_dir: str
    selected_episodes: list[int]
    completed_episodes: list[int]
    current_episode: Optional[int] = None
    current_progress: Optional[DownloadProgress] = None

    def save(self, path: str):
        """Save state to JSON file"""
        with open(path, 'w', encoding='utf-8') as f:
            json.dump(asdict(self), f, ensure_ascii=False, indent=2)

    @classmethod
    def load(cls, path: str) -> Optional['DownloadState']:
        """Load state from JSON file"""
        try:
            with open(path, 'r', encoding='utf-8') as f:
                data = json.load(f)
                if data.get('current_progress'):
                    data['current_progress'] = DownloadProgress(**data['current_progress'])
                return cls(**data)
        except (FileNotFoundError, json.JSONDecodeError):
            return None


class VideoDownloader:
    """Downloads videos with resume support"""

    CHUNK_SIZE = 1024 * 1024  # 1MB chunks

    def __init__(self, output_dir: str = "./output"):
        self.output_dir = Path(output_dir)
        self.output_dir.mkdir(parents=True, exist_ok=True)
        self.state_file = self.output_dir / "download_state.json"
        self.state: Optional[DownloadState] = None
        self._paused = False
        self._cancelled = False

        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
            'Accept': '*/*',
            'Accept-Encoding': 'identity',
        })

    def pause(self):
        """Pause download"""
        self._paused = True

    def resume(self):
        """Resume download"""
        self._paused = False

    def cancel(self):
        """Cancel download"""
        self._cancelled = True

    def is_paused(self) -> bool:
        return self._paused

    def is_cancelled(self) -> bool:
        return self._cancelled

    def reset(self):
        """Reset pause/cancel state"""
        self._paused = False
        self._cancelled = False

    def init_state(self, series_info: SeriesInfo, selected_episodes: list[int]):
        """Initialize download state"""
        self.state = DownloadState(
            series_id=series_info.series_id,
            series_title=series_info.title,
            total_episodes=series_info.total_episodes,
            output_dir=str(self.output_dir),
            selected_episodes=sorted(selected_episodes),
            completed_episodes=[],
            current_episode=None,
            current_progress=None
        )
        self._save_state()

    def load_state(self) -> Optional[DownloadState]:
        """Load existing download state"""
        self.state = DownloadState.load(str(self.state_file))
        return self.state

    def _save_state(self):
        """Save current state"""
        if self.state:
            self.state.save(str(self.state_file))

    def get_episode_filename(self, episode: int) -> Path:
        """Get output filename for episode"""
        return self.output_dir / f"ep_{episode:02d}.mp4"

    def get_remaining_episodes(self) -> list[int]:
        """Get list of episodes still to download"""
        if not self.state:
            return []
        return [ep for ep in self.state.selected_episodes
                if ep not in self.state.completed_episodes]

    def download_episode(
        self,
        episode_info: EpisodeInfo,
        progress_callback: Optional[Callable[[int, int, float], None]] = None,
        use_tqdm: bool = True
    ) -> bool:
        """
        Download a single episode with resume support

        Args:
            episode_info: Episode information with video URL
            progress_callback: Optional callback(downloaded, total, speed)
            use_tqdm: Whether to show tqdm progress bar

        Returns:
            True if download completed successfully
        """
        output_file = self.get_episode_filename(episode_info.episode_number)
        temp_file = output_file.with_suffix('.mp4.part')

        # Check for existing partial download
        downloaded_bytes = 0
        if temp_file.exists():
            downloaded_bytes = temp_file.stat().st_size

        # Update state
        if self.state:
            self.state.current_episode = episode_info.episode_number

        try:
            # Make initial request to get file size
            headers = {}
            if downloaded_bytes > 0:
                headers['Range'] = f'bytes={downloaded_bytes}-'

            response = self.session.get(
                episode_info.video_url,
                headers=headers,
                stream=True,
                timeout=30
            )

            # Handle range response
            if response.status_code == 206:
                # Partial content - resuming
                content_range = response.headers.get('Content-Range', '')
                total_bytes = int(content_range.split('/')[-1]) if '/' in content_range else 0
            elif response.status_code == 200:
                # Full content - starting fresh
                total_bytes = int(response.headers.get('Content-Length', 0))
                downloaded_bytes = 0  # Server doesn't support resume
            else:
                print(f"Error: HTTP {response.status_code}")
                return False

            # Update state with progress
            if self.state:
                self.state.current_progress = DownloadProgress(
                    episode=episode_info.episode_number,
                    downloaded_bytes=downloaded_bytes,
                    total_bytes=total_bytes,
                    completed=False
                )
                self._save_state()

            # Setup progress bar
            pbar = None
            if use_tqdm:
                pbar = tqdm(
                    total=total_bytes,
                    initial=downloaded_bytes,
                    unit='B',
                    unit_scale=True,
                    desc=f"ตอนที่ {episode_info.episode_number}"
                )

            # Download with resume
            mode = 'ab' if downloaded_bytes > 0 and response.status_code == 206 else 'wb'
            start_time = time.time()
            bytes_since_last_update = 0

            with open(temp_file, mode) as f:
                for chunk in response.iter_content(chunk_size=self.CHUNK_SIZE):
                    # Check for pause/cancel
                    while self._paused and not self._cancelled:
                        time.sleep(0.1)

                    if self._cancelled:
                        if pbar:
                            pbar.close()
                        self._save_state()
                        return False

                    if chunk:
                        f.write(chunk)
                        downloaded_bytes += len(chunk)
                        bytes_since_last_update += len(chunk)

                        if pbar:
                            pbar.update(len(chunk))

                        # Calculate speed and call callback
                        if progress_callback:
                            elapsed = time.time() - start_time
                            speed = bytes_since_last_update / elapsed if elapsed > 0 else 0
                            progress_callback(downloaded_bytes, total_bytes, speed)

                        # Update state periodically
                        if self.state and bytes_since_last_update >= self.CHUNK_SIZE * 5:
                            self.state.current_progress = DownloadProgress(
                                episode=episode_info.episode_number,
                                downloaded_bytes=downloaded_bytes,
                                total_bytes=total_bytes,
                                completed=False
                            )
                            self._save_state()
                            bytes_since_last_update = 0
                            start_time = time.time()

            if pbar:
                pbar.close()

            # Rename temp file to final (use replace for Windows atomic overwrite support)
            temp_file.replace(output_file)

            # Mark as completed
            if self.state:
                if episode_info.episode_number not in self.state.completed_episodes:
                    self.state.completed_episodes.append(episode_info.episode_number)
                self.state.current_episode = None
                self.state.current_progress = None
                self._save_state()

            return True

        except requests.RequestException as e:
            print(f"Download error: {e}")
            self._save_state()
            return False

    def download_all(
        self,
        episodes: list[EpisodeInfo],
        progress_callback: Optional[Callable[[int, int, int, int], None]] = None
    ) -> tuple[int, int]:
        """
        Download multiple episodes

        Args:
            episodes: List of episode info
            progress_callback: Optional callback(current_ep, total_eps, downloaded, total_bytes)

        Returns:
            Tuple of (successful_count, failed_count)
        """
        successful = 0
        failed = 0

        for i, episode in enumerate(episodes):
            if self._cancelled:
                break

            # Skip already completed
            if self.state and episode.episode_number in self.state.completed_episodes:
                successful += 1
                continue

            print(f"\nDownloading episode {episode.episode_number}...")

            if self.download_episode(episode):
                successful += 1
            else:
                if not self._cancelled:
                    failed += 1

            if progress_callback:
                progress_callback(i + 1, len(episodes), 0, 0)

        return successful, failed

    def clear_state(self):
        """Clear download state file"""
        if self.state_file.exists():
            self.state_file.unlink()
        self.state = None


# Test
if __name__ == "__main__":
    from parser import RongyokParser

    parser = RongyokParser()
    downloader = VideoDownloader("./output")

    # Get series info
    series_info = parser.get_series_info(941)
    if series_info:
        print(f"Series: {series_info.title}")
        print(f"Episodes: {series_info.total_episodes}")

        # Initialize for first 2 episodes
        downloader.init_state(series_info, [1, 2])

        # Get episode URLs
        ep1 = parser.get_episode_video_url(941, 1)
        if ep1:
            print(f"\nDownloading episode 1...")
            success = downloader.download_episode(ep1)
            print(f"Success: {success}")

File: src-tauri/src/downloader.rs
use crate::parser::RongyokParser;
use crate::state::save_state;
use crate::types::{DownloadProgressEvent, DownloadState, EpisodeInfo, LogMessage};
use futures_util::StreamExt;
use reqwest::header::{RANGE, USER_AGENT};
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Instant;
use tauri::{AppHandle, Emitter};
use tokio::fs::OpenOptions;
use tokio::io::AsyncWriteExt;


#[derive(Clone)]
pub struct VideoDownloader {
    parser: RongyokParser,
    is_paused: Arc<AtomicBool>,
    is_cancelled: Arc<AtomicBool>,
}

impl VideoDownloader {
    pub fn new(parser: RongyokParser) -> Self {
        Self {
            parser,
            is_paused: Arc::new(AtomicBool::new(false)),
            is_cancelled: Arc::new(AtomicBool::new(false)),
        }
    }

    pub fn pause(&self) {
        self.is_paused.store(true, Ordering::SeqCst);
    }

    pub fn resume(&self) {
        self.is_paused.store(false, Ordering::SeqCst);
    }

    pub fn cancel(&self) {
        self.is_cancelled.store(true, Ordering::SeqCst);
    }

    pub fn reset_controls(&self) {
        self.is_paused.store(false, Ordering::SeqCst);
        self.is_cancelled.store(false, Ordering::SeqCst);
    }

    pub fn is_cancelled(&self) -> bool {
        self.is_cancelled.load(Ordering::SeqCst)
    }

    pub fn get_episode_filename(output_dir: &str, episode: u32) -> PathBuf {
        Path::new(output_dir).join(format!("ep_{:02}.mp4", episode))
    }

    pub async fn download_episode(
        &self,
        episode_info: &EpisodeInfo,
        output_dir: &str,
        total_episodes_in_batch: u32,
        current_batch_index: u32,
        app: &AppHandle,
    ) -> Result<bool, String> {
        let output_file = Self::get_episode_filename(output_dir, episode_info.episode_number);
        let temp_file = Path::new(output_dir).join(format!("ep_{:02}.mp4.part", episode_info.episode_number));

        // Create output directory if needed
        tokio::fs::create_dir_all(output_dir)
            .await
            .map_err(|e| format!("Failed to create output directory: {}", e))?;

        // Check for existing partial download
        let mut downloaded_bytes: u64 = 0;
        if temp_file.exists() {
            if let Ok(meta) = tokio::fs::metadata(&temp_file).await {
                downloaded_bytes = meta.len();
            }
        }

        let client = reqwest::Client::new();
        let mut req = client
            .get(&episode_info.video_url)
            .header(
                USER_AGENT,
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
            )
            .header("Referer", "https://rongyok.com/")
            .header("Accept", "*/*")
            .header("Sec-Fetch-Dest", "video")
            .header("Sec-Fetch-Mode", "no-cors")
            .header("Sec-Fetch-Site", "cross-site");

        if downloaded_bytes > 0 {
            req = req.header(RANGE, format!("bytes={}-", downloaded_bytes));
        }

        let response = req
            .send()
            .await
            .map_err(|e| format!("Network request failed: {}", e))?;

        let status = response.status();
        let total_bytes = if status == reqwest::StatusCode::PARTIAL_CONTENT {
            if let Some(cr) = response.headers().get("content-range").and_then(|h| h.to_str().ok()) {
                if let Some(total_str) = cr.split('/').last() {
                    total_str.parse::<u64>().unwrap_or(0)
                } else {
                    0
                }
            } else {
                0
            }
        } else if status == reqwest::StatusCode::OK {
            downloaded_bytes = 0; // Server doesn't support range
            response.content_length().unwrap_or(0)
        } else {
            return Err(format!("Server returned HTTP status {}", status));
        };

        // Open temp file
        let mut file = if downloaded_bytes > 0 && status == reqwest::StatusCode::PARTIAL_CONTENT {
            OpenOptions::new()
                .create(true)
                .append(true)
                .open(&temp_file)
                .await
                .map_err(|e| format!("Failed to open temp file for appending: {}", e))?
        } else {
            OpenOptions::new()
                .create(true)
                .write(true)
                .truncate(true)
                .open(&temp_file)
                .await
                .map_err(|e| format!("Failed to create temp file: {}", e))?
        };

        let mut stream = response.bytes_stream();
        let mut start_time = Instant::now();
        let mut bytes_since_last_tick: u64 = 0;
        let mut last_progress_emit = Instant::now();

        while let Some(chunk_result) = stream.next().await {
            // Check pause / cancel
            while self.is_paused.load(Ordering::SeqCst) && !self.is_cancelled.load(Ordering::SeqCst) {
                tokio::time::sleep(tokio::time::Duration::from_millis(150)).await;
            }

            if self.is_cancelled.load(Ordering::SeqCst) {
                return Ok(false);
            }

            let chunk = chunk_result.map_err(|e| format!("Error streaming chunk: {}", e))?;
            file.write_all(&chunk)
                .await
                .map_err(|e| format!("Failed to write to file: {}", e))?;

            downloaded_bytes += chunk.len() as u64;
            bytes_since_last_tick += chunk.len() as u64;

            // Emit progress event every 300ms
            if last_progress_emit.elapsed().as_millis() >= 300 {
                let elapsed_secs = start_time.elapsed().as_secs_f64();
                let speed_bytes_per_sec = if elapsed_secs > 0.0 {
                    bytes_since_last_tick as f64 / elapsed_secs
                } else {
                    0.0
                };

                let speed_formatted = format!("{:.2} MB/s", speed_bytes_per_sec / (1024.0 * 1024.0));

                let remaining_bytes = if total_bytes > downloaded_bytes {
                    total_bytes - downloaded_bytes
                } else {
                    0
                };

                let eta_formatted = if speed_bytes_per_sec > 0.0 && remaining_bytes > 0 {
                    let eta_secs = (remaining_bytes as f64 / speed_bytes_per_sec) as u64;
                    let mins = eta_secs / 60;
                    let secs = eta_secs % 60;
                    format!("{:02}:{:02}", mins, secs)
                } else {
                    "--:--".to_string()
                };

                let percentage = if total_bytes > 0 {
                    (downloaded_bytes as f64 / total_bytes as f64) * 100.0
                } else {
                    0.0
                };

                let _ = app.emit(
                    "download-progress",
                    DownloadProgressEvent {
                        episode: episode_info.episode_number,
                        total_episodes: total_episodes_in_batch,
                        current_episode_index: current_batch_index,
                        downloaded_bytes,
                        total_bytes,
                        percentage,
                        speed_bytes_per_sec,
                        speed_formatted,
                        eta_formatted,
                        status_message: format!(
                            "Downloading Episode {} ({:.1}%)",
                            episode_info.episode_number, percentage
                        ),
                    },
                );

                bytes_since_last_tick = 0;
                start_time = Instant::now();
                last_progress_emit = Instant::now();
            }
        }

        file.flush()
            .await
            .map_err(|e| format!("Failed to flush file: {}", e))?;
        drop(file);

        // Atomic replace temp file -> output file
        if output_file.exists() {
            let _ = tokio::fs::remove_file(&output_file).await;
        }

        tokio::fs::rename(&temp_file, &output_file)
            .await
            .map_err(|e| format!("Failed to rename temp file to output file: {}", e))?;

        Ok(true)
    }

    pub async fn download_batch(
        &self,
        series_id: u32,
        series_title: String,
        episodes: Vec<u32>,
        output_dir: String,
        auto_merge: bool,
        delete_parts: bool,
        app: AppHandle,
    ) -> Result<(), String> {
        self.reset_controls();

        let mut completed_episodes = Vec::new();
        let total_eps = episodes.len() as u32;

        let mut state = DownloadState {
            series_id,
            series_title: series_title.clone(),
            total_episodes: total_eps,
            output_dir: output_dir.clone(),
            selected_episodes: episodes.clone(),
            completed_episodes: Vec::new(),
            current_episode: None,
            current_progress: None,
        };

        save_state(&state, &output_dir)?;

        for (idx, &ep_num) in episodes.iter().enumerate() {
            if self.is_cancelled() {
                break;
            }

            state.current_episode = Some(ep_num);
            save_state(&state, &output_dir)?;

            let _ = app.emit(
                "log-message",
                LogMessage {
                    id: format!("{}-ep-{}", series_id, ep_num),
                    timestamp: chrono::Local::now().format("%H:%M:%S").to_string(),
                    level: "info".to_string(),
                    text: format!("Resolving dynamic stream for Episode {}...", ep_num),
                },
            );

            // Fetch dynamic stream URL
            let ep_info = match self.parser.get_episode_video_url(series_id, ep_num).await {
                Ok(info) => info,
                Err(e) => {
                    let _ = app.emit(
                        "log-message",
                        LogMessage {
                            id: format!("{}-err-{}", series_id, ep_num),
                            timestamp: chrono::Local::now().format("%H:%M:%S").to_string(),
                            level: "error".to_string(),
                            text: format!("Failed to get stream for Episode {}: {}", ep_num, e),
                        },
                    );
                    continue;
                }
            };

            let _ = app.emit(
                "log-message",
                LogMessage {
                    id: format!("{}-start-{}", series_id, ep_num),
                    timestamp: chrono::Local::now().format("%H:%M:%S").to_string(),
                    level: "info".to_string(),
                    text: format!("Starting download for Episode {}...", ep_num),
                },
            );

            match self
                .download_episode(&ep_info, &output_dir, total_eps, (idx + 1) as u32, &app)
                .await
            {
                Ok(true) => {
                    completed_episodes.push(ep_num);
                    state.completed_episodes = completed_episodes.clone();
                    save_state(&state, &output_dir)?;

                    let _ = app.emit(
                        "log-message",
                        LogMessage {
                            id: format!("{}-done-{}", series_id, ep_num),
                            timestamp: chrono::Local::now().format("%H:%M:%S").to_string(),
                            level: "success".to_string(),
                            text: format!("Episode {} download completed successfully.", ep_num),
                        },
                    );
                }
                Ok(false) => {
                    let _ = app.emit(
                        "log-message",
                        LogMessage {
                            id: format!("{}-cancel-{}", series_id, ep_num),
                            timestamp: chrono::Local::now().format("%H:%M:%S").to_string(),
                            level: "warning".to_string(),
                            text: format!("Download cancelled on Episode {}.", ep_num),
                        },
                    );
                    break;
                }
                Err(e) => {
                    let _ = app.emit(
                        "log-message",
                        LogMessage {
                            id: format!("{}-fail-{}", series_id, ep_num),
                            timestamp: chrono::Local::now().format("%H:%M:%S").to_string(),
                            level: "error".to_string(),
                            text: format!("Episode {} failed: {}", ep_num, e),
                        },
                    );
                }
            }
        }

        // Auto Merge if requested
        if auto_merge && !self.is_cancelled() && completed_episodes.len() >= 2 {
            let _ = app.emit(
                "log-message",
                LogMessage {
                    id: format!("{}-merge-start", series_id),
                    timestamp: chrono::Local::now().format("%H:%M:%S").to_string(),
                    level: "info".to_string(),
                    text: format!("Merging {} downloaded episodes into unified MP4...", completed_episodes.len()),
                },
            );

            let merger = crate::merger::VideoMerger::new();
            if merger.is_available() {
                let mut video_files = Vec::new();
                for &ep in &completed_episodes {
                    let path = Self::get_episode_filename(&output_dir, ep);
                    if path.exists() {
                        video_files.push(path);
                    }
                }

                // Sanitize title for filename
                let clean_title = series_title
                    .replace(['<', '>', ':', '"', '/', '\\', '|', '?', '*'], "")
                    .trim()
                    .to_string();

                let out_merged = Path::new(&output_dir).join(format!("{}.mp4", clean_title));

                match merger.merge_videos(&video_files, &out_merged, delete_parts).await {
                    Ok(_) => {
                        let _ = app.emit(
                            "log-message",
                            LogMessage {
                                id: format!("{}-merge-success", series_id),
                                timestamp: chrono::Local::now().format("%H:%M:%S").to_string(),
                                level: "success".to_string(),
                                text: format!("Video merged successfully: {:?}", out_merged),
                            },
                        );
                    }
                    Err(e) => {
                        let _ = app.emit(
                            "log-message",
                            LogMessage {
                                id: format!("{}-merge-fail", series_id),
                                timestamp: chrono::Local::now().format("%H:%M:%S").to_string(),
                                level: "error".to_string(),
                                text: format!("FFmpeg merge error: {}", e),
                            },
                        );
                    }
                }
            }
        }

        state.current_episode = None;
        save_state(&state, &output_dir)?;

        Ok(())
    }
}
